xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,721 @@
+      requestWriter.onError(error);
+    }
+
+    private void cleanUp() {
+      if (adsStream == this) {
+        adsStream = null;
+      }
+    }
+
+    private void sendLdsRequest(String resourceName) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(ldsVersion)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(ADS_TYPE_URL_LDS)
+              .setResponseNonce(ldsRespNonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    private void sendRdsRequest(String resourceName) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      rdsResourceName = resourceName;
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(rdsVersion)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(ADS_TYPE_URL_RDS)
+              .setResponseNonce(rdsRespNonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    /**
+     * Prepares a DiscoveryRequest containing the given information as an ACK.
+     */
+    private void prepareAckRequest(String typeUrl, String resourceName, String versionInfo,
+        String nonce) {
+      pendingAckRequest =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .build();
+    }
+
+    /**
+     * Sends the previously prepared ACK DiscoveryRequest to management server. Updates the latest
+     * accepted version for the corresponding resource type.
+     */
+    private void sendPendingAckRequest() {
+      checkState(pendingAckRequest != null, "There is no pending ACK request");
+      checkState(requestWriter != null, "ADS stream has not been started");
+      // Update the latest accepted version.
+      String typeUrl = pendingAckRequest.getTypeUrl();
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        ldsVersion = pendingAckRequest.getVersionInfo();
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        rdsVersion = pendingAckRequest.getVersionInfo();
+      }
+      // TODO: cases for (VHDS/)CDS/EDS.
+      requestWriter.onNext(pendingAckRequest);
+      pendingAckRequest = null;
+    }
+
+    /**
+     * Cancels the pending ACK request and sends its corresponding (with version reverted)
+     * NACK request.
+     */
+    private void nackPendingAckRequest() {
+      checkState(pendingAckRequest != null, "There is no pending ACK request");
+      checkState(requestWriter != null, "ADS stream has not been started");
+      DiscoveryRequest.Builder nackRequestBuilder = pendingAckRequest.toBuilder();
```

Comment:
Although not very clear the [text](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-versioning)

> If Envoy had instead rejected configuration update X, it would reply with error_detail populated and its previous version, which in this case was the empty initial version.

 implies that error_detail needs to be populated in your Nack request. Here you are not populating it (need to call `DiscoveryRequest.Builder.setErrorDetail()`). I think you should have some error to send in each case of Nack

Reply:
Our current design did not mention this, we don't have a spec for this yet. So not implemented.

Reply:
Update: added some sort of error messages into `error_detail` to help TD debugging.